File: serve_h5.js
const http = require('http');
const httpProxy = require('http-proxy');
const fs = require('fs');
const path = require('path');

const STATIC_DIR = '/workspace/h5build/dist/dev/h5';
const BACKEND = 'http://127.0.0.1:8091';
const PORT = process.env.PORT || 8080;

const proxy = httpProxy.createProxyServer({});

const MIME_TYPES = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'application/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.svg': 'image/svg+xml',
  '.ico': 'image/x-icon',
  '.woff': 'font/woff',
  '.woff2': 'font/woff2',
  '.ttf': 'font/ttf',
  '.map': 'application/json',
  '.mp3': 'audio/mpeg',
  '.mp4': 'video/mp4',
};

const server = http.createServer((req, res) => {
  const url = new URL(req.url, `http://${req.headers.host}`);
  const pathname = url.pathname;

  // API proxy
  if (pathname.startsWith('/api/') || pathname.startsWith('/push')) {
    proxy.web(req, res, { target: BACKEND }, (err, target, proxyRes) => {
      if (err) {
        console.error(`Proxy error: ${err.message}`);
        if (!res.headersSent) {
          res.writeHead(502, { 'Content-Type': 'application/json' });
          res.end(JSON.stringify({ code: 502, message: 'Backend unavailable' }));
        }
      }
    });
    return;
  }

  // Static files
  let filePath = path.join(STATIC_DIR, pathname);
  
  // Security check - prevent directory traversal
  if (!filePath.startsWith(STATIC_DIR)) {
    res.writeHead(403);
    res.end('Forbidden');
    return;
  }

  // Default to index.html
  if (pathname === '/' || pathname === '') {
    filePath = path.join(STATIC_DIR, 'index.html');
  }

  fs.stat(filePath, (err, stats) => {
    if (err || !stats.isFile()) {
      // SPA fallback for unknown routes
      filePath = path.join(STATIC_DIR, 'index.html');
    }

    const ext = path.extname(filePath).toLowerCase();
    const contentType = MIME_TYPES[ext] || 'application/octet-stream';

    fs.readFile(filePath, (err, data) => {
      if (err) {
        res.writeHead(500);
        res.end('Internal Server Error');
        return;
      }
      res.writeHead(200, {
        'Content-Type': contentType,
        'Cache-Control': ext === '.html' ? 'no-cache' : 'public, max-age=31536000',
      });
      res.end(data);
    });
  });
});

server.listen(PORT, '0.0.0.0', () => {
  console.log(`H5 server running on http://0.0.0.0:${PORT}`);
  console.log(`Static files: ${STATIC_DIR}`);
  console.log(`API proxy: ${BACKEND}`);
});

File: serve_h5.py
import http.server
import socketserver
import urllib.request
import urllib.error
import os
import sys

BACKEND = 'http://127.0.0.1:8091'
STATIC_DIR = '/workspace/h5build/dist/dev/h5'
PORT = 8080

class ProxyHandler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=STATIC_DIR, **kwargs)
    
    def do_GET(self):
        if self.path.startswith('/api/') or self.path.startswith('/push'):
            self.proxy_request()
        elif self.path in ('/', '/index.html'):
            # 返回 SPA 入口
            return super().do_GET()
        else:
            # 检查是否存在静态文件
            filepath = os.path.join(STATIC_DIR, self.path.lstrip('/'))
            if os.path.isfile(filepath):
                return super().do_GET()
            else:
                # SPA fallback
                self.path = '/index.html'
                return super().do_GET()
    
    def do_POST(self):
        if self.path.startswith('/api/') or self.path.startswith('/push'):
            self.proxy_request()
        else:
            self.send_error(405)
    
    def proxy_request(self):
        # 转发请求到后端
        url = BACKEND + self.path
        content_length = int(self.headers.get('Content-Length', 0))
        body = self.rfile.read(content_length) if content_length > 0 else None
        
        headers = {}
        for key in self.headers:
            if key.lower() not in ('host', 'connection', 'transfer-encoding'):
                headers[key] = self.headers[key]
        headers['Host'] = '127.0.0.1:8091'
        
        req = urllib.request.Request(
            url,
            data=body,
            headers=headers,
            method=self.command
        )
        
        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                self.send_response(resp.status)
                for key, val in resp.headers.items():
                    if key.lower() not in ('connection', 'transfer-encoding'):
                        self.send_header(key, val)
                self.end_headers()
                self.wfile.write(resp.read())
        except urllib.error.HTTPError as e:
            self.send_response(e.code)
            for key, val in e.headers.items():
                if key.lower() not in ('connection', 'transfer-encoding'):
                    self.send_header(key, val)
            self.end_headers()
            self.wfile.write(e.read())
        except Exception as e:
            self.send_error(502, str(e))

class ReusableTCPServer(socketserver.TCPServer):
    allow_reuse_address = True
    request_queue_size = 128

if __name__ == '__main__':
    PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 8080
    with ReusableTCPServer(('0.0.0.0', PORT), ProxyHandler) as httpd:
        print(f'H5 server running on http://0.0.0.0:{PORT}')
        print(f'Static files: {STATIC_DIR}')
        print(f'API proxy: {BACKEND}')
        httpd.serve_forever()
